### PR TITLE
Update Objective-C API

### PR DESCRIPTION
**Pull request by edgchen1**
**Description**
Update Objective-C API to be more usable from Swift. E.g., to allow conversion from Objective-C methods with trailing NSError** parameter to throwing Swift methods.
Update CMake Objective-C framework setup. There's probably more work to be done here.

**Motivation and Context**
Use ORT from Objective-C and Swift.

### CODE REVIEW

File: objectivec/include/ort_value.h [hunk 1]
```diff
@@ -3,37 +3,25 @@
...
+- (instancetype)init NS_UNAVAILABLE;
 
 /**
  * Creates a value that is a tensor.
  * The tensor data is allocated by the caller.
  *
- * @param data The tensor data.
+ * @param tensorData The tensor data.
  * @param type The tensor data element type.
  * @param shape The tensor shape.
```
Comment (skottmckay) [inline objectivec/include/ort_value.h:26]: nit: elementType

File: objectivec/src/ort_enums.mm [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+#include <algorithm>
+
+#include "core/session/onnxruntime_cxx_api.h"
+
+namespace {
+struct LoggingLevelInfo {
+  ORTLoggingLevel logging_level;
+  OrtLoggingLevel capi_logging_level;
+};
```
Comment (skottmckay) [inline objectivec/src/ort_enums.mm:14]: Are there any other options for disambiguating ORTLoggingLevel and OrtLoggingLevel?

File: objectivec/src/ort_enums.mm [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+// supported ORT tensor element data types
+// define the mapping from ORTTensorElementDataType to C API
+// ONNXTensorElementDataType here
+constexpr TensorElementTypeInfo kElementTypeInfos[]{
+    {ORTTensorElementDataTypeUndefined, ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED, 0},
+    {ORTTensorElementDataTypeFloat, ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT, sizeof(float)},
+    {ORTTensorElementDataTypeInt32, ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32, sizeof(int32_t)},
+};
```
Comment (skottmckay) [inline objectivec/src/ort_enums.mm:51]: Is it possible to expand this list in the near future (e.g. support 8-bit input for quantized models)?

Comment (edgchen1) [reply objectivec/src/ort_enums.mm:51]: yes, it is

File: objectivec/src/ort_value.mm [hunk 4]
```diff
@@ -3,90 +3,58 @@
...
+    const Ort::TypeInfo& CXXAPITypeInfo) {
+  auto* result = [[ORTValueTypeInfo alloc] init];
+  const auto valueType = CXXAPITypeInfo.GetONNXType();
+
+  result.type = CAPIToPublicValueType(valueType);
+
+  if (valueType == ONNX_TYPE_TENSOR) {
+    const auto tensorTypeAndShapeInfo =
+        CXXAPITypeInfo.GetTensorTypeAndShapeInfo();
+    result.tensorTypeAndShapeInfo =
```
Comment (skottmckay) [inline objectivec/src/ort_value.mm:45]: nit: can we use the same 120 char line limit we use everywhere else?

**Review by skottmckay — APPROVED** [review]

File: objectivec/src/ort_enums.mm (hunk 2 above)
Comment (edgchen1) [reply objectivec/src/ort_enums.mm:14]: none that come to mind now

**Review by skottmckay — APPROVED** [review]

**Comment by wenbingl** [comment]
Overall, are all these header compatible with Swift bridging header file, or you need create the the bridging header file differently?

It's important to make sure the interfaces can be reused in Swift, which is mainstream for Apple developer, I think.

**Comment by edgchen1** [comment]
I created a test Swift program with a bridging header with the contents below and was able to use the types from Swift.
```
#import "onnxruntime.h"
```

---
In reply to: [832835610](https://github.com/microsoft/onnxruntime/pull/7567#issuecomment-832835610) [](http://example.com/codeflow?ancestors=832835610,832835610)

**Comment by wenbingl** [comment]
> I created a test Swift program with a bridging header with the contents below and was able to use the types from Swift.
> 
> ```
> #import "onnxruntime.h"
> ```
> 
> In reply to: [832835610](https://github.com/microsoft/onnxruntime/pull/7567#issuecomment-832835610)

Cool, in that case, should these bridging header could be put in this folder as well? Ideally, or was separated in the swift package?

**Comment by edgchen1** [comment]
hm, I think the bridging header contents would depend on the user's project. for example, they might need to import other Objective-C headers in addition to ours.
the packaging is not finalized yet, we can potentially include one if that's useful.

---
In reply to: [832892169](https://github.com/microsoft/onnxruntime/pull/7567#issuecomment-832892169) [](http://example.com/codeflow?ancestors=832892169,832892169)

**Pull request merged**